File: day1/main.ts
import { strict as assert } from "assert";
import { readFileSync } from "fs";

const solve1 = (file: string): number => {
  const nums: number[] = readFileSync(file, "utf-8")
    .trim()
    .split("\n")
    .map((num) => +num);
  return nums.reduce((a, b) => a + b, 0);
};

const solve2 = function (file: string): number {
  const nums: number[] = readFileSync(file, "utf-8")
    .trim()
    .split("\n")
    .map((num) => +num);

  const seen: Set<number> = new Set<number>();
  let sum = 0;
  let found = false;

  while (!found) {
    for (const num of nums) {
      seen.add(sum);
      sum += num;

      if (seen.has(sum)) {
        found = true;
        break;
      }
    }
  }

  return sum;
};

assert(solve1("./example.txt") === 3);
console.log(solve1("./input.txt"));

assert(solve2("./example2.txt") === 2);
console.log(solve2("./input.txt"));

File: day1/src/main.rs
use std::{
    collections::HashSet,
    fs::File,
    io::{prelude::*, BufReader},
    path::Path,
};

fn solve1(filename: &str) -> i64 {
    lines_from_file(filename)
        .iter()
        .map(|s| s.parse::<i64>().unwrap())
        .sum()
}

fn solve2(filename: &str) -> i64 {
    let nums: Vec<i64> = lines_from_file(filename)
        .iter()
        .map(|s| s.parse::<i64>().unwrap())
        .collect();

    let mut seen = HashSet::<i64>::new();
    let mut sum: i64 = 0;
    let mut found = false;

    while !found {
        for num in &nums {
            seen.insert(sum);
            sum += num;

            if seen.contains(&sum) {
                found = true;
                break;
            }
        }
    }

    sum
}

fn lines_from_file(filename: impl AsRef<Path>) -> Vec<String> {
    let file = File::open(filename).expect("no such file");
    let buf = BufReader::new(file);
    buf.lines()
        .map(|l| l.expect("Could not parse line"))
        .collect()
}

fn main() {
    println!("{}", solve1("../input/2018/day1.txt"));
    println!("{}", solve2("../input/2018/day1.txt"));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        assert_eq!(solve1("example.txt"), 3);
        assert_eq!(solve2("example2.txt"), 2);
    }
}
